pkg/controller/redis/redis_controller.go

```diff
@@ -0,0 +1,177 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller_redis")
+
+// Add creates a new Redis Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	logger := logrus.WithFields(logrus.Fields{"controller": "controller_redis"})
+	return &ReconcileRedis{client: mgr.GetClient(), scheme: mgr.GetScheme(), logger: logger}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("redis-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource Redis
+	err = c.Watch(&source.Kind{Type: &integreatlyv1alpha1.Redis{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// TODO(user): Modify this to be the types you create that are owned by the primary resource
+	// Watch for changes to secondary resource Pods and requeue the owner Redis
+	err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &integreatlyv1alpha1.Redis{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileRedis implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileRedis{}
+
+// ReconcileRedis reconciles a Redis object
+type ReconcileRedis struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	logger *logrus.Entry
+}
+
+// Reconcile reads that state of the cluster for a Redis object and makes changes based on the state read
+// and what is in the Redis.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileRedis) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.TODO()
+	providerList := []providers.RedisProvider{aws.NewAWSRedisProvider(r.client, r.logger), openshift.NewOpenShiftRedisProvider(r.client, r.logger)}
```

Comment:
do we want to do it this way? Generally would avoid creating new objects like this within a function. I would recommend creating the objects outside and passing them in as dependencies.

Reply:
We can move this logic to `newReconciler`, and make the internal reconcile function the public function. We would have been testing `reconcile` here

Reply:
:+1: